src/main/java/com/commercetools/paymenttoorderprocessor/jobs/actions/MessageReader.java

```diff
@@ -43,51 +44,72 @@
     @Value("${ctp.messagereader.minutesoverlapping}")
     private Integer minutesOverlapping;
 
-    private List<PaymentTransactionStateChangedMessage> messages = Collections.emptyList();
+    @Nonnull
+    private Queue<PaymentTransactionStateChangedMessage> unprocessedMessagesQueue = new ArrayDeque<>();
+
     private boolean wasInitialQueried = false;
     private long total;
     private long offset = 0L;
-    private final int RESULTSPERPAGE = 500;
-    private final int PAGEOVERLAP = 5;
-
-    private MessageQuery messageQuery;
+    final int RESULTSPERPAGE = 500;
+    final int PAGEOVERLAP = 5;
 
+    /**
+     * @return the oldest unprocessed message from the queue, or <b>null</b> if no new messages to process in CTP.
+     */
     @Override
+    @Nullable
     public PaymentTransactionStateChangedMessage read() {
         LOG.debug("wasInitialQueried: {}", wasInitialQueried);
-        if (isQueryNeeded()) {
-            getUnprocessedMessagesFromPlatform();
+
+        while (isQueryNeeded()) {
+            fetchUnprocessedMessagesFromPlatform();
         }
-        return getMessageFromList();
+
+        return getUnprocessedMessageFromQueue();
     }
```

Comment:
I think this code would be easier to read , when fetchUnprocessedMessagesFromPlatform return the List of messages, and this List would be passed to  fetchUnprocessedMessagesFromPlatform

Reply:
no, i exactly removed the List: if we work with queue, it is better to use Queue.

Reply:
also, this kind of recursion (e.g. return list, and then pass this list to the same function) is even more confusing (IMHO)